unit_tests/log/test_main_chkpt_info.cpp

```diff
@@ -0,0 +1,960 @@
+{
+  test_env_chkpt env {220, 80};
+
+  cubpacking::packer pac;
+  size_t size = env.get_before ()->get_packed_size (pac, 0);
+  size += env.get_before ()->get_packed_size (pac, size);
+  char buffer[size];
+  pac.set_buffer (buffer, size);
+  env.get_before ()->pack (pac);
+  env.get_before ()->pack (pac);
+
+  cubpacking::unpacker unpac;
+  unpac.set_buffer (buffer, size);
+  env.get_after ()->unpack (unpac);
+  env.require_equal (*env.get_before (), *env.get_after ());
+
+  checkpoint_info after_2;
+
+  after_2.unpack (unpac);
+
+  env.require_equal (*env.get_after (), after_2);
+}
+
+std::map<TRANID, log_tdes *> tran_map;
+std::map<TRANID, log_tdes *> systb_System_tdes;
+int count_2pc_after_recovery;
+
+TEST_CASE ("Test load and recovery on empty tran table", "")
+{
+  test_env_chkpt env;
+  LOG_LSA smallest_lsa;
+  LOG_LSA star_lsa;
+  THREAD_ENTRY thd;
+
+  env.get_after ()->load_trantable_snapshot (&thd, smallest_lsa);
+  env.get_after ()->recovery_analysis (&thd, star_lsa);
+  env.get_after ()->recovery_2pc_analysis (&thd);
+  REQUIRE (tran_map.size() == 0);
+}
+
+int
+search_for_id (TRANID id)
+{
+  for (int i = 0; i < log_Gl.trantable.num_total_indices; i++)
+    {
+      if (log_Gl.trantable.all_tdes[i]->trid == id)
+	{
+	  return 1;
+	}
+    }
+  return 0;
+}
+
+void
+full_compare_tdes (log_tdes *td1, const log_tdes *td2)
+{
+  REQUIRE (td1->head_lsa == td2->head_lsa);
+  REQUIRE (td1->tail_lsa == td2->tail_lsa);
+  REQUIRE (td1->undo_nxlsa == td2->undo_nxlsa);
+  REQUIRE (td1->posp_nxlsa == td2->posp_nxlsa);
+  REQUIRE (td1->savept_lsa == td2->savept_lsa);
+  REQUIRE (td1->tail_topresult_lsa == td2->tail_topresult_lsa);
+  REQUIRE (td1->rcv.sysop_start_postpone_lsa == td2->rcv.sysop_start_postpone_lsa);
+  REQUIRE (td1->rcv.atomic_sysop_start_lsa == td2->rcv.atomic_sysop_start_lsa);
+}
+
+void
+full_compare_tdes_active (log_tdes *td1, const log_tdes *td2)
+{
+  REQUIRE (td1->head_lsa == td2->head_lsa);
+  REQUIRE (td1->tail_lsa == td2->undo_nxlsa);
+  REQUIRE (td1->tail_lsa == td2->tail_lsa);
+  REQUIRE (td1->posp_nxlsa == td2->posp_nxlsa);
+  REQUIRE (td1->savept_lsa == td2->savept_lsa);
+  REQUIRE (td1->tail_topresult_lsa == td2->tail_topresult_lsa);
+  REQUIRE (td1->rcv.sysop_start_postpone_lsa == td2->rcv.sysop_start_postpone_lsa);
+  REQUIRE (td1->rcv.atomic_sysop_start_lsa == td2->rcv.atomic_sysop_start_lsa);
+}
+
+void
+check_recovery (checkpoint_info obj)
+{
+  std::map<TRANID, log_tdes *>::iterator itr;
+  for (itr = tran_map.begin(); itr != tran_map.end(); ++itr)
+    {
+      REQUIRE (itr->first != NULL_TRANID);
+    }
+
+  for (int i = 1; i < log_Gl.trantable.num_total_indices; i++)
+    {
+      LOG_TDES *tdes = log_Gl.trantable.all_tdes[i];
+
+      if (log_Gl.trantable.all_tdes[i]->trid == NULL_TRANID)
+	{
+	  continue;
+	}
+
+      std::map<TRANID, log_tdes *>::const_iterator tdes_after = tran_map.find (tdes->trid);
+      if (tdes->trid != NULL_TRANID && tdes->tail_lsa == NULL_LSA)
```

Comment:
tdes->trid == NULL_TRANID is already skipped at line 213.
```suggestion
      if (tdes->tail_lsa == NULL_LSA)
```